lms/services/async_oauth_http.py

```diff
@@ -0,0 +1,60 @@
+import asyncio
+from typing import List
+
+import aiohttp
+
+from lms.services import ExternalRequestError
+
+
+class AsyncOAuthHTTPService:
+    def __init__(self, oauth2_token_service):
+        self._oauth2_token_service = oauth2_token_service
+
+    def get(
+        self, urls, concurrency=10, headers=None, timeout=10
+    ) -> List[aiohttp.ClientResponse]:
+        """
+        Sends async requests to all urls.
+
+        :raise OAuth2TokenError: if we don't have an access token for the user
+        :raise ExternalRequestError: if something goes wrong with the HTTP
+            request
+        """
+
+        headers = headers or {}
+
+        access_token = self._oauth2_token_service.get().access_token
+        headers["Authorization"] = f"Bearer {access_token}"
+
+        async def _async_request(semaphore, aio_session, url, headers=None):
+            try:
+                async with semaphore:
+                    async with aio_session.get(
+                        url, headers=headers, timeout=timeout
+                    ) as response:
+                        return response
+            except aiohttp.ClientError as err:
+                raise ExternalRequestError("Blackboard async request failed") from err
+
+        async def _prepare_requests(urls, headers=None):
+            semaphore = asyncio.Semaphore(concurrency)
+
+            async with aiohttp.ClientSession() as session:
+                tasks = []
+                for url in urls:
+                    task = asyncio.create_task(
+                        _async_request(semaphore, session, url, headers)
+                    )
+                    tasks.append(task)
+                try:
+                    return await asyncio.gather(*tasks, return_exceptions=False)
+                except ExternalRequestError:
+                    for task in tasks:
+                        task.cancel()
+                    raise
+
+        return asyncio.run(_prepare_requests(urls, headers))
```

Comment:
I was worrying about the ordering here but it's guaranteed by asyncio.gather:

https://docs.python.org/3/library/asyncio-task.html#asyncio.gather

```The order of result values corresponds to the order of awaitables in aws.```